## Day 48 Lecture 2 Assignment

In this assignment, we will apply density-based clustering to a dataset containing the locations of all Starbucks in the U.S.

This assignment will also use the haversine and plotly packages, which you should already have installed from the previous assignment.

In [1]:
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.cluster import DBSCAN
import plotly.express as px

This dataset contains the latitude and longitude (as well as several other details we will not be using) of every Starbucks in the world as of February 2017. Each row consists of the following features, which are generally self-explanatory:

- Brand
- Store Number
- Store Name
- Ownership Type
- Street Address
- City
- State/Province
- Country
- Postcode
- Phone Number
- Timezone
- Longitude
- Latitude

Load in the dataset.

In [5]:
# answer goes here
df = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/Data%20Sets%20Clustering/starbucks_locations.csv')
df.head()

Brand  Store Number  ... Longitude Latitude
0  Starbucks  47370-257954  ...      1.53    42.51
1  Starbucks  22331-212325  ...     55.47    25.42
2  Starbucks  47089-256771  ...     55.47    25.39
3  Starbucks  22126-218024  ...     54.38    24.48
4  Starbucks  17127-178586  ...     54.54    24.51

[5 rows x 13 columns]

Begin by narrowing down the dataset to a specific geographic area of interest. Try just the United States; since you won't be calculating a distance matrix you can use more than just one state.

In [8]:
# answer goes here
sb = df.loc[df.Country == 'US']
sb.head()

Brand  Store Number  ... Longitude Latitude
11964  Starbucks   3513-125945  ...   -149.78    61.21
11965  Starbucks   74352-84449  ...   -149.84    61.14
11966  Starbucks  12449-152385  ...   -149.85    61.11
11967  Starbucks  24936-233524  ...   -149.89    61.13
11968  Starbucks    8973-85630  ...   -149.86    61.14

[5 rows x 13 columns]

Build a DBSCAN clustering model using eps=2 (miles) and min_samples=5. Some tips that may be helpful:

1. Unlike our approach for hierarchical clustering, we do not need to calculate the NxN distance matrix for DBSCAN upfront. It directly supports the haversine distance metric, provided the nearest-neighbors algorithm is a ball tree. Set the "algorithm" and "metric" parameters to the appropriate values. 
2. Scikit-learn's implementation of haversine distance expects radians instead of degrees. Therefore, it would be advisable to create two new columns, Lat_Rad and Lon_Rad, that convert the Latitude and Longitude columns into radians. (Hint: there is a numpy function that does this.)  
3. The eps parameter, which corresponds to the radius of the neighborhood, will also need to be in radians. The conversion factor for miles to radians is approximately 1/3958.748; in other words, if you want the neighborhood to have a radius of 3 miles, set eps = 3/3958.748.  

Side note: ball-tree is an indexing structure that is very useful for nearest-neighbor calculations. The general time-complexity of finding a nearest neighbor using a Ball Tree is O(nlog(n)). This is a vast improvement over the naive O($n^{2}$) and allows us to cluster on much larger subsets of the data, like the entire country. Scikit-learn directly supports creating ball-trees through sklearn.neighbors.BallTree; if inclined, you could extend the analysis in the first after-lecture assignment (in which we calculated a similarity matrix for Hawaii) to the entire country using a BallTree and identify "island Starbucks locations" on a much larger scale.

Additionally, save the predicted cluster assignments as a new column in your dataframe.

In [ ]:
# converting Lat and Lon to radians to new columns
sb["Lat_Rad"] = np.radians(sb.Latitude)
sb["Lon_Rad"] = np.radians(sb.Longitude)

In [13]:
sb.head()

Brand  Store Number  ...   Lat_Rad   Lon_Rad
11964  Starbucks   3513-125945  ...  1.068316 -2.614154
11965  Starbucks   74352-84449  ...  1.067094 -2.615201
11966  Starbucks  12449-152385  ...  1.066571 -2.615376
11967  Starbucks  24936-233524  ...  1.066920 -2.616074
11968  Starbucks    8973-85630  ...  1.067094 -2.615550

[5 rows x 15 columns]

In [21]:
# model prep
scan = DBSCAN(metric="haversine", eps=2/3958.748, min_samples=5, algorithm="ball_tree")
sb["db_cluster"] = scan.fit_predict(sb[["Lat_Rad", "Lon_Rad"]])
sb["db_cluster"].value_counts()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



-1      6881
 31      345
 26      336
 265     307
 397     225
        ... 
 159       4
 297       4
 280       4
 106       4
 416       3
Name: db_cluster, Length: 426, dtype: int64

Finally, plot the resulting clusters on a map using the "scatter_geo" function from plotly.express. The map defaults to the entire world; the "scope" parameter is useful for narrowing down the region plotted in the map. The documentation can be found here:

https://www.plotly.express/plotly_express/#plotly_express.scatter_geo

How many clusters did DBSCAN produce? How many locations were treated as outliers (cluster = -1)?

*There are 6881 outliers with 2 miles being the radius and min_samples=5.*

In [24]:
# answer goes here
px.scatter_geo(lat=sb.Latitude, lon=sb.Longitude, color=sb["db_cluster"], scope='usa')

From the previous plot, we should see a very large number of clusters (400+). This would suggest that our definition of neighborhood may have been too strict. Experiment with other values of eps and min_samples and see how your changes affect the output. Output a map with what you think is the "best" clustering result below.

In [47]:
# answer goes here
scan = DBSCAN(metric="haversine", eps=50/3958.748, min_samples=5, algorithm="ball_tree")
sb["db_cluster"] = scan.fit_predict(sb[["Lat_Rad", "Lon_Rad"]])
sb["db_cluster"].value_counts()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



14    4311
9     2853
56     967
2      933
15     653
      ... 
40       5
24       5
39       5
66       5
63       5
Name: db_cluster, Length: 72, dtype: int64

In [48]:
px.scatter_geo(lat=sb.Latitude, lon=sb.Longitude, color=sb["db_cluster"], scope='usa')

*With the epsilon distance set at 50 miles, the outliers are well-defined especially on the midwest.*